lib/core/src/Cardano/DB/Sqlite.hs

```diff
@@ -345,19 +360,38 @@ instance MatchMigrationError SqliteException where
 newtype ManualMigration = ManualMigration
     { executeManualMigration :: Sqlite.Connection -> IO () }
 
-createSqliteBackend
+createSqlitePool
     :: Tracer IO DBLog
     -> Maybe FilePath
     -> ManualMigration
     -> LogFunc
-    -> IO (SqlBackend, Sqlite.Connection)
-createSqliteBackend trace fp migration logFunc = do
+    -> IO (Pool (SqlBackend, Sqlite.Connection))
+createSqlitePool tr fp migration logFunc = do
     let connStr = sqliteConnStr fp
-    traceWith trace $ MsgConnStr connStr
-    conn <- Sqlite.open connStr
-    executeManualMigration migration conn
-    backend <- wrapConnectionInfo (mkSqliteConnectionInfo connStr) conn logFunc
-    pure (backend, conn)
+    traceWith tr $ MsgConnStr connStr
+
+    let createConnection = do
+            let info = mkSqliteConnectionInfo connStr
+            conn <- Sqlite.open connStr
+            executeManualMigration migration conn
+            backend <- wrapConnectionInfo info conn logFunc
+            pure (backend, conn)
+
+    let destroyConnection = \(backend, _) -> do
+            destroySqliteBackend tr backend fp
+
+    createPool
+        createConnection
+        destroyConnection
+        numberOfStripes
+        timeToLive
+        maximumConnections
+  where
+    numberOfStripes = 1
+    timeToLive = 600 :: NominalDiffTime
+    -- When running in :memory:, we want a single connection that does not get
+    -- cleaned up.
+    maximumConnections = maybe 1 (const 10) fp
```

Comment:
Is there some difference with sqlite in-memory databases which mean only one concurrent connection is possible? Otherwise, maximumConnections may as well be the same for both.

The number 10 seems fine for each wallet database. Perhaps in future, to prevent scaling bottlenecks, the maximum for the stake pools database could be scaled by the number of wallets.

Reply:
I originally wanted to use [createSqlitePool](https://hackage.haskell.org/package/persistent-sqlite-2.10.6.2/docs/Database-Persist-Sqlite.html#v:createSqlitePool) from `persistent` but it doesn't give any ways to access the underlying connection (only the persistent abstraction `SqlBackend`) which we do need in a few places for "hardcore error handling" and also custom migrations. 

The `createSqlitePool` function rightfully mentions:

> Note that this should not be used with the :memory: connection string, as the pool will regularly remove connections, destroying your database. Instead, use withSqliteConn.

I stole the default `10` max connections and `600` seconds time to live from the default chosen in persistent. Using `1` for in the in-memory database seems to work-around the issue nicely as resource-pool always keep at least one connection. Since this is only used in testing, I found it to be a simple and non-invasive resolution.